### ISSUE TITLE
[tune] Population-Based Training Focuses on the Finding a Good Trained Model Rather Than a Set of Good Hyperparameters

### ISSUE THREAD

**Issue by leimao** [ray-project/ray#9178]
### What is your question?
I personally think the objective of Ray-Tune is a little bit misleading. At the end of population-based training, Ray-Tune will tell us what the best configuration is for the hyperparameters. However, given the nature of population-based training, if the user leaves Ray-Tune and directly apply the "best hyperparameters" Ray-Tune suggested for training the model, the training will not be reproducible. So the goal of poplulation-based training is to help you train a better model using genetically mutated hyperparameters during training, rather than finding a good set of "fixed hyperparameters".

To reproduce the trained model, we would need the entire trajectory of hyperparameter mutations of the best model during training. Is there any simple interface in Ray-Tune to reproduce the training of the best model?

Thank you. 

*Ray version and other system information (Python version, TensorFlow version, OS):*
Ray 0.9.0dev
Python 3.6

**Comment by richardliaw** [comment]
This issue is a repeat of #https://github.com/ray-project/ray/issues/7307. 

As a workaround, you can currently recover the trace of hyperparameters for each trained model by viewing the PBT logs. See: https://docs.ray.io/en/master/tune/api_docs/schedulers.html#population-based-training-tune-schedulers-populationbasedtraining.

**Comment by leimao** [comment]
@richardliaw Thank you. I checked the json file for a dummy run. It looks like as follows:
```
{"score": -0.31925569020162725, "done": false, "timesteps_total": null, "episodes_total": null, "training_iteration": 1, "experiment_id": "0076871752bf46b9b7ce2b1e95020436", "date": "2020-06-29_15-45-44", "timestamp": 1593445544, "time_this_iter_s": 5.7220458984375e-06, "time_total_s": 5.7220458984375e-06, "pid": 2973, "hostname": "a3bbe166e4a1", "node_ip": "172.17.0.2", "config": {"h_0": 0.548858622576932, "h_1": 0.7152178474357822}, "time_since_restore": 5.7220458984375e-06, "timesteps_since_restore": 0, "iterations_since_restore": 1, "trial_id": "97e2f_00000", "experiment_tag": "0_h_0=0.54886,h_1=0.71522"}
{"score": -0.22488156503352252, "done": false, "timesteps_total": null, "episodes_total": null, "training_iteration": 2, "experiment_id": "0076871752bf46b9b7ce2b1e95020436", "date": "2020-06-29_15-45-44", "timestamp": 1593445544, "time_this_iter_s": 1.33514404296875e-05, "time_total_s": 1.9073486328125e-05, "pid": 2973, "hostname": "a3bbe166e4a1", "node_ip": "172.17.0.2", "config": {"h_0": 0.548858622576932, "h_1": 0.7152178474357822}, "time_since_restore": 1.9073486328125e-05, "timesteps_since_restore": 0, "iterations_since_restore": 2, "trial_id": "97e2f_00000", "experiment_tag": "0_h_0=0.54886,h_1=0.71522"}
{"score": -0.136468402929254, "done": false, "timesteps_total": null, "episodes_total": null, "training_iteration": 3, "experiment_id": "0076871752bf46b9b7ce2b1e95020436", "date": "2020-06-29_15-45-44", "timestamp": 1593445544, "time_this_iter_s": 6.198883056640625e-06, "time_total_s": 2.5272369384765625e-05, "pid": 2973, "hostname": "a3bbe166e4a1", "node_ip": "172.17.0.2", "config": {"h_0": 0.548858622576932, "h_1": 0.7152178474357822}, "time_since_restore": 2.5272369384765625e-05, "timesteps_since_restore": 0, "iterations_since_restore": 3, "trial_id": "97e2f_00000", "experiment_tag": "0_h_0=0.54886,h_1=0.71522"}
{"score": -0.05363365727017011, "done": false, "timesteps_total": null, "episodes_total": null, "training_iteration": 4, "experiment_id": "0076871752bf46b9b7ce2b1e95020436", "date": "2020-06-29_15-45-44", "timestamp": 1593445544, "time_this_iter_s": 5.245208740234375e-06, "time_total_s": 3.0517578125e-05, "pid": 2973, "hostname": "a3bbe166e4a1", "node_ip": "172.17.0.2", "config": {"h_0": 0.548858622576932, "h_1": 0.7152178474357822}, "time_since_restore": 3.0517578125e-05, "timesteps_since_restore": 0, "iterations_since_restore": 4, "trial_id": "97e2f_00000", "experiment_tag": "0_h_0=0.54886,h_1=0.71522"}
{"score": 0.023980305287767845, "done": true, "timesteps_total": null, "episodes_total": null, "training_iteration": 5, "experiment_id": "0076871752bf46b9b7ce2b1e95020436", "date": "2020-06-29_15-45-44", "timestamp": 1593445544, "time_this_iter_s": 9.298324584960938e-06, "time_total_s": 3.981590270996094e-05, "pid": 2973, "hostname": "a3bbe166e4a1", "node_ip": "172.17.0.2", "config": {"h_0": 0.548858622576932, "h_1": 0.7152178474357822}, "time_since_restore": 3.981590270996094e-05, "timesteps_since_restore": 0, "iterations_since_restore": 5, "trial_id": "97e2f_00000", "experiment_tag": "0_h_0=0.54886,h_1=0.71522"}
```
It does contain all the information required to reproduce the training. So in principle, the user could write a text parser and apply the hyperparameters for each training iteration in their training script, and the model trained using the training script should be exactly the same to the one trained using Ray-Tune. However, I am still a little bit worried about the randomness which affects reproducibility. Say in their training script and the Ray-Tune script, they both use random seed 0, but the training result might be slightly different due to that the training script and Ray-Tune have their own unique components that might affect the randomness. **It would be very desirable that Ray-Tune could import this json file and reproduce the training from scratch.** This might not be a priority, but it should be a good feature to have.

Best,

Lei

**Comment by krfricke** [comment]
We added a replay utility in #9953
